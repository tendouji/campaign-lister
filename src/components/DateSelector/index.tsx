import React from 'react';
import { connect } from 'react-redux';
import styled from "styled-components";
import DatePicker from "react-datepicker";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCalendarCheck } from '@fortawesome/free-solid-svg-icons/faCalendarCheck'
import { faArrowsAltH } from '@fortawesome/free-solid-svg-icons/faArrowsAltH'
import { faTimes } from "@fortawesome/free-solid-svg-icons/faTimes";
import { colors, layout, styles } from '../../constants/ui';
import { setSnackBarDisplay, setDateRange } from "../../actions";
import { DateRangeData, ReducerStateData, SnackBarData } from '../../models';

import "react-datepicker/dist/react-datepicker.css";


type DateSelectorProps = {
    [key:string]: any
}

export class DateSelector extends React.Component<DateSelectorProps, any> {
    constructor(props: DateSelectorProps) {
        super(props);

        this.setDate = this.setDate.bind(this);
        this.onClearDateRange = this.onClearDateRange.bind(this);
        this.onDateFilterClick = this.onDateFilterClick.bind(this);

        const dateNow: Date = new Date();

        this.state = {
            dateNow,
            startDate: dateNow,
            endDate: dateNow,
        };
    }

    componentDidUpdate(prevProps: any) {
        const { appState } = this.props;
        const { dateRange } = appState;

        if(prevProps.appState.dateRange.startDate !== dateRange.startDate) {
            if(!dateRange.startDate) {
                // NOTE: startDate is null
                this.onClearDateRange();
            }
        }
    }

    onDateFilterClick(e?: React.MouseEvent | undefined) {
        const { startDate, endDate } = this.state;

        if(startDate > endDate) {
            const { setSnackBarDisplay } = this.props;
            setSnackBarDisplay({
                show: true,
                message: 'Start date cannot be later than end date.'
            });
            return;
        }

        const { setDateRange } = this.props;
        setDateRange({
            startDate,
            endDate,
        });
    };

    setDate(dateVal: Date, type: string) {
        this.setState({ [type]: dateVal }, () => {
            const { startDate, endDate } = this.state;
            if(startDate > endDate) {
                this.setState({ endDate: startDate }, () => {
                    this.onDateFilterClick();
                });
            } else {
                this.onDateFilterClick();
            }
        });
    };

    onClearDateRange() {
        const { dateNow } = this.state;
        const { setDateRange } = this.props;
        setDateRange({ startDate: null, endDate: null });
        this.setState({
            startDate: dateNow,
            endDate: dateNow
        });
    };

    render() {
        const { startDate, endDate } = this.state;
        const { appState } = this.props;
        const { dateRange } = appState;

        return (
            <DateSelectorWrapper className="date-selector">
                <div className="date-area">
                    <div className="date-col start-date">
                        <div className="label">Start Date</div>
                        <DatePicker
                            id="startDatePicker"
                            selected={startDate}
                            onChange={(date: Date) => this.setDate(date, 'startDate')}
                            placeholderText='Select your start date'
                            popperPlacement="bottom"
                            todayButton="Today"
                            className="datepicker-input"
                        />
                    </div>
                    <div className="icon-range">
                        <FontAwesomeIcon icon={faArrowsAltH} />
                    </div>
                    <div className="date-col end-date">
                        <div className="label">End Date</div>
                        <DatePicker
                            id="endDatePicker"
                            selected={endDate < startDate ? startDate : endDate}
                            onChange={(date: Date) => this.setDate(date, 'endDate')}
                            minDate={startDate}
                            placeholderText='Select your end date'
                            popperPlacement="bottom"
                            todayButton="Today"
                            className="datepicker-input"
                        />
                    </div>
                </div>
                { !!dateRange.startDate || !!dateRange.endDate
                    ? <button id="clearDateRange" onClick={this.onClearDateRange}><FontAwesomeIcon icon={faTimes} /></button>
                    : <button id="goFilter" onClick={this.onDateFilterClick}><FontAwesomeIcon icon={faCalendarCheck} /></button>
                }
            </DateSelectorWrapper>
        );
    }
}

const mapStateToProps = (state: ReducerStateData) => {
    return ({ appState: state.appState })
};

const mapDispatchToProps = (dispatch: any) => ({
    setDateRange: (data: DateRangeData) => dispatch(setDateRange(data)),
    setSnackBarDisplay: (data: SnackBarData) => dispatch(setSnackBarDisplay(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(DateSelector);



const DateSelectorWrapper = styled.div`
    display: flex;
    width: 100%;
    align-items: center;
    
    & .date-area {
        display: flex;
        width: calc(100% - ${layout.filterBarHeight});
        padding: 0 ${layout.largeGap};
        box-sizing: border-box;
        align-items: center;
        justify-content: space-between;
    }
    
    & .date-col {
        box-sizing: border-box;
        font-size: ${layout.fontSizeStandard};
        
        & > .label {
            margin-bottom: ${layout.xSmallGap};
            font-size: ${layout.fontSizeSmall};
            font-weight: 500;
        }
    }
    
    & .icon-range {
        padding: 1.4rem 3px 0;
    }
    
    & > button {
        ${styles.filterBarButtonStyle};
    }
    
    & .react-datepicker__triangle {
        left: 50%;
    }
    
    & .react-datepicker__input-container {
        & > input[type="text"] {
            width: 6rem;
            padding: ${layout.xSmallGap} ${layout.smallGap};
            border: ${styles.lineStyle};
            font-size: ${layout.fontSizeStandard};
            text-align: center;
            background-color: ${colors.lightGray};
            box-sizing: border-box;
            
            &:focus {
                outline: none;
            }
        }
    }
`;