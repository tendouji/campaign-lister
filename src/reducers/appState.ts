import {
    SET_SEARCH_KEY,
    SET_DATE_RANGE,
    SET_SNACKBAR_DISPLAY,
    SET_MODALPOPUP_DISPLAY,
} from '../constants/actionType';
import { ActionData, AppStateData } from "../models";


const defaultAppState: AppStateData = {
    isLoading: false,
    searchKey: '',
    dateRange: {
        startDate: null,
        endDate: null,
    },
    snackBarDisplay: {
        show: false,
        message: ''
    },
    modalPopupDisplay: {
        show: false,
        title: '',
        message: ''
    }
};

export const appStateReducer = (state = defaultAppState, action: ActionData) => {
    switch (action.type) {
        case SET_SEARCH_KEY:
            return {
                ...state,
                searchKey: action.data
            };
        case SET_DATE_RANGE:
            return {
                ...state,
                dateRange: {
                    ...state.dateRange,
                    ...action.data
                }
            };
        case SET_SNACKBAR_DISPLAY:
            return {
                ...state,
                snackBarDisplay: {
                    ...state.snackBarDisplay,
                    ...action.data
                }
            };
        case SET_MODALPOPUP_DISPLAY:
            return {
                ...state,
                modalPopupDisplay: {
                    ...state.modalPopupDisplay,
                    ...action.data
                }
            };
        default:
            return state;
    }
};