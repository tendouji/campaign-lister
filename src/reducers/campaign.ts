import { STORE_CAMPAIGN_DATA } from '../constants/actionType';
import { ActionData, CampaignData } from "../models";


const defaultCampaignState: CampaignData[] = [];

export const campaignReducer = (state = defaultCampaignState, action: ActionData) => {
    switch (action.type) {
        case STORE_CAMPAIGN_DATA:
            return [
                ...state,
                ...action.data
            ];
        default:
            return state;
    }
};