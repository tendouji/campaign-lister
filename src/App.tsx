import React from 'react';
import { connect } from "react-redux";
import {
    appTitle,
    helpTitleString,
    noHelpContent
} from "./constants/text";
import {
    CampaignData,
    DateRangeData, ModalPopupData,
    ReducerStateData,
    SnackBarData
} from "./models";
import {
    setDateRange, setModalPopupDisplay,
    setSearchKey,
    setSnackBarDisplay,
    storeCampaignData
} from "./actions";
import { Header } from "./components/Header";
import { FilterBar } from "./components/FilterBar";
import SnackBar from "./components/SnackBar";
import CampaignList from "./components/CampaignList";
import { Footer } from './components/Footer';
import ModalPopup from "./components/ModalPopup";


declare global {
    interface Window {
        EventQueue: any;
        AddCampaignsEventName: string;
        GetHelp: any;
    }
}

export class App extends React.Component<any, any> {
    constructor(props: any) {
        super(props);

        this.storeCampaigns = this.storeCampaigns.bind(this);
    }

    componentDidMount() {
        if(!!window && !!window.EventQueue) {
            window.EventQueue.subscribe(window.AddCampaignsEventName, this.storeCampaigns);
        }
    }

    storeCampaigns(data: CampaignData[]) {
        console.log(`${appTitle}: Store campaigns data into redux`);

        const {
            storeCampaignData,
            setSearchKey,
            setDateRange
        } = this.props;

        setSearchKey('');
        setDateRange({ startDate: null, endDate: null });
        storeCampaignData(data);
    };

    closeSnackBar = () => {
        const { setSnackBarDisplay } = this.props;
        setSnackBarDisplay({ show: false });
    };

    closeModalPopup = () => {
        const { setModalPopupDisplay } = this.props;
        setModalPopupDisplay({ show: false });
    };

    render() {
        const { appState, campaigns } = this.props;
        const { snackBarDisplay, modalPopupDisplay } = appState;

        return (
            <>
                <Header title={appTitle} />
                <FilterBar disabled={!campaigns || campaigns.length === 0} />
                <CampaignList />
                <Footer />

                <ModalPopup
                    show={modalPopupDisplay.show}
                    title={helpTitleString}
                    content={!!window.GetHelp && typeof window.GetHelp === 'function'
                        ? window.GetHelp()
                        : noHelpContent
                    }
                    close={this.closeModalPopup}
                />

                <SnackBar
                    show={snackBarDisplay.show}
                    message={snackBarDisplay.message}
                    close={this.closeSnackBar}
                />
            </>
        );
    }
};


const mapStateToProps = (state: ReducerStateData) => {
    return ({
        appState: state.appState,
        campaigns: state.campaigns
    })
};

const mapDispatchToProps = (dispatch: any) => ({
    setSearchKey: (data: string) => dispatch(setSearchKey(data)),
    setDateRange: (data: DateRangeData) => dispatch(setDateRange(data)),
    storeCampaignData: (data: CampaignData[]) => dispatch(storeCampaignData(data)),
    setSnackBarDisplay: (data: SnackBarData) => dispatch(setSnackBarDisplay(data)),
    setModalPopupDisplay: (data: ModalPopupData) => dispatch(setModalPopupDisplay(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(App);

