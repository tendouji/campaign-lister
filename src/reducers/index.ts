import { combineReducers } from 'redux';
import { campaignReducer } from './campaign';
import { appStateReducer } from './appState';


export const appReducer = combineReducers({
    appState: appStateReducer,
    campaigns: campaignReducer,
});