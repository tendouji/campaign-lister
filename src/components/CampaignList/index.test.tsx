import React from 'react';
import { createStore } from "redux";
import { mount, shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import { appReducer } from "../../reducers";
import { CampaignList } from './';
import {DateSelector} from "../DateSelector";
import {App} from "../../App";


describe('<DateSelector />', () => {
    const store = createStore(appReducer);
    const mockData = [
        {"id":1,"name":"Campaign AAA 1","startDate":"1/1/2018","endDate":"12/31/2018","budget":999999},
        {"id":2,"name":"Campaign AAA 2","startDate":"1/1/2019","endDate":"12/31/2020", "budget":999999},
        {"id":3,"name":"Campaign BBB","startDate":"1/1/2019","endDate":"12/31/2020","budget":999999},
    ];
    const mockSearchKey = 'AAA';
    const mockStartDate = new Date('1/1/2019');
    const mockEndDate = new Date('12/31/2019');

    const props = {
        store,
        campaigns: mockData,
        appState: {
            searchKey: '',
            dateRange: { }
        }
    };

    it('renders without crashing', () => {
        const wrapper = mount(<CampaignList {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('filter campaigns correctly based on searchKey', () => {
        const wrapper = mount(<CampaignList {...props} />);
        const instance = (wrapper.instance() as CampaignList);

        const result = instance.filterCampaignsBySearchKey(mockSearchKey);
        expect(result.length).toEqual(2);
    });

    it('filter campaigns correctly based on dateRange', () => {
        const wrapper = mount(<CampaignList {...props} />);
        const instance = (wrapper.instance() as CampaignList);

        const result = instance.filterCampaignsByDateRange(mockStartDate, mockEndDate);
        expect(result.length).toEqual(2);
    });

    it('filter campaigns correctly based on searchKey and dateRange', () => {
        const wrapper = mount(<CampaignList {...props} />);

        const instance = (wrapper.instance() as CampaignList);
        const dateRangeResult = instance.filterCampaignsByDateRange(mockStartDate, mockEndDate);
        wrapper.setProps({
            ...props,
            appState: {
                ...props.appState,
                dateRange: {
                    startDate: mockStartDate,
                    endDate: mockEndDate
                }
            }
        }, () => {
            wrapper.setState({ filteredCampaigns: dateRangeResult }, () => {
                const result = instance.filterCampaignsBySearchKey('AAA');
                expect(result.length).toEqual(1);
            });
        });
    });
});