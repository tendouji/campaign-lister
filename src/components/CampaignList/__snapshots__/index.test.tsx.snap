// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateSelector /> renders without crashing 1`] = `
<CampaignList
  appState={
    Object {
      "dateRange": Object {},
      "searchKey": "",
    }
  }
  campaigns={
    Array [
      Object {
        "budget": 999999,
        "endDate": "12/31/2018",
        "id": 1,
        "name": "Campaign AAA 1",
        "startDate": "1/1/2018",
      },
      Object {
        "budget": 999999,
        "endDate": "12/31/2020",
        "id": 2,
        "name": "Campaign AAA 2",
        "startDate": "1/1/2019",
      },
      Object {
        "budget": 999999,
        "endDate": "12/31/2020",
        "id": 3,
        "name": "Campaign BBB",
        "startDate": "1/1/2019",
      },
    ]
  }
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <styled.div
    className="campaign-header"
  >
    <StyledComponent
      className="campaign-header"
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": false,
            "lastClassName": "kdbIGY",
            "rules": Array [
              "
    display: flex;
    width: 100%;
    padding: ",
              "1rem",
              ";
    box-sizing: border-box;
    justify-content: space-between;
    align-items: center;
    
    & .title {
        width: calc(100% - ",
              "10rem",
              ");
        
        & > h2 {
            margin: 0 0 ",
              "0.625rem",
              ";
        }
        
        & .description {
            & > span {
                font-weight: 700;
                color: ",
              "#43cadb",
              ";
            }
        }
    }
    
    & .legend {
        width: ",
              "10rem",
              ";
        padding: ",
              "0.625rem",
              ";
        border: ",
              "1px solid #ddd",
              ";
        box-sizing: border-box;
        text-align: center;
        font-size: ",
              "0.75rem",
              ";
        font-weight: 500;
        color: ",
              "#666",
              ";
        
        & .legend-title {
            margin-bottom: ",
              "0.3125rem",
              ";
            font-weight: 700;
            text-transform: uppercase;
        }
        
        & .legend-row {
            display: flex;
            justify-content: space-between;
            
            & .col {
                & .label { }
            }
        }
    }
    
    @media screen and (max-width: ",
              "37.5rem",
              ") {
        display: block;
        padding: ",
              "1rem",
              " ",
              "1rem",
              " ",
              "0.625rem",
              ";
    
        & .title {
            width: 100%;
            margin-bottom: ",
              "1.25rem",
              ";
        }
        
        & .legend {
            display: flex;
            margin: 0 auto;
            width: 14rem;
        
            & .legend-title {
                width: 3rem;
                margin-bottom: 0;
                margin-right: ",
              "0.3125rem",
              ";
            }
            
            & .legend-row {
                display: flex;
                width: calc(100% - 3rem);
                justify-content: space-between;
            }
        }
    }
    
    @media screen and (max-width: ",
              "25rem",
              ") {        
        & .legend {
            display: none;
        }
    }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-bwzfXH kdbIGY campaign-header"
      >
        <div
          className="title"
        >
          <h2>
            Campaign list
          </h2>
          <div
            className="description"
            dangerouslySetInnerHTML={
              Object {
                "__html": "Results for <span>all</span> campaigns (0 entry)",
              }
            }
          />
        </div>
        <div
          className="legend"
        >
          <div
            className="legend-title"
          >
            Status Type
          </div>
          <div
            className="legend-row"
          >
            <div
              className="col"
            >
              <div
                className="label"
              >
                ongoing
              </div>
              <StatusDot
                status="ongoing"
              >
                <styled.span
                  className="status-dot ongoing"
                >
                  <StyledComponent
                    className="status-dot ongoing"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bdVaJa",
                          "isStatic": false,
                          "lastClassName": "GNxtY",
                          "rules": Array [
                            "
    & > span { 
        display: none;
    }

    &.",
                            "ongoing",
                            " {
        color: ",
                            "#4caf50",
                            ";
    }
    
    &.",
                            "ended",
                            " {
        color: ",
                            "#ff2717",
                            ";
    }
    
    &.",
                            "upcoming",
                            " {
        color: ",
                            "#ffc107",
                            ";
    }
    
    @media screen and (max-width: ",
                            "25rem",
                            ") {
        & > span { 
            display: inline-block;
            margin-left: ",
                            "0.3125rem",
                            ";
            line-height: 1rem;
        }
    }    
",
                          ],
                        },
                        "displayName": "styled.span",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bdVaJa",
                        "target": "span",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <span
                      className="sc-bdVaJa GNxtY status-dot ongoing"
                    >
                      <FontAwesomeIcon
                        border={false}
                        className=""
                        fixedWidth={false}
                        flip={null}
                        icon={
                          Object {
                            "icon": Array [
                              512,
                              512,
                              Array [],
                              "f192",
                              "M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm80 248c0 44.112-35.888 80-80 80s-80-35.888-80-80 35.888-80 80-80 80 35.888 80 80z",
                            ],
                            "iconName": "dot-circle",
                            "prefix": "fas",
                          }
                        }
                        inverse={false}
                        listItem={false}
                        mask={null}
                        pull={null}
                        pulse={false}
                        rotation={null}
                        size={null}
                        spin={false}
                        swapOpacity={false}
                        symbol={false}
                        title=""
                        transform={null}
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-dot-circle fa-w-16 "
                          data-icon="dot-circle"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          style={Object {}}
                          viewBox="0 0 512 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm80 248c0 44.112-35.888 80-80 80s-80-35.888-80-80 35.888-80 80-80 80 35.888 80 80z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </FontAwesomeIcon>
                      <span>
                        ongoing
                      </span>
                    </span>
                  </StyledComponent>
                </styled.span>
              </StatusDot>
            </div>
            <div
              className="col"
            >
              <div
                className="label"
              >
                ended
              </div>
              <StatusDot
                status="ended"
              >
                <styled.span
                  className="status-dot ended"
                >
                  <StyledComponent
                    className="status-dot ended"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bdVaJa",
                          "isStatic": false,
                          "lastClassName": "GNxtY",
                          "rules": Array [
                            "
    & > span { 
        display: none;
    }

    &.",
                            "ongoing",
                            " {
        color: ",
                            "#4caf50",
                            ";
    }
    
    &.",
                            "ended",
                            " {
        color: ",
                            "#ff2717",
                            ";
    }
    
    &.",
                            "upcoming",
                            " {
        color: ",
                            "#ffc107",
                            ";
    }
    
    @media screen and (max-width: ",
                            "25rem",
                            ") {
        & > span { 
            display: inline-block;
            margin-left: ",
                            "0.3125rem",
                            ";
            line-height: 1rem;
        }
    }    
",
                          ],
                        },
                        "displayName": "styled.span",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bdVaJa",
                        "target": "span",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <span
                      className="sc-bdVaJa GNxtY status-dot ended"
                    >
                      <FontAwesomeIcon
                        border={false}
                        className=""
                        fixedWidth={false}
                        flip={null}
                        icon={
                          Object {
                            "icon": Array [
                              512,
                              512,
                              Array [],
                              "f192",
                              "M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm80 248c0 44.112-35.888 80-80 80s-80-35.888-80-80 35.888-80 80-80 80 35.888 80 80z",
                            ],
                            "iconName": "dot-circle",
                            "prefix": "fas",
                          }
                        }
                        inverse={false}
                        listItem={false}
                        mask={null}
                        pull={null}
                        pulse={false}
                        rotation={null}
                        size={null}
                        spin={false}
                        swapOpacity={false}
                        symbol={false}
                        title=""
                        transform={null}
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-dot-circle fa-w-16 "
                          data-icon="dot-circle"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          style={Object {}}
                          viewBox="0 0 512 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm80 248c0 44.112-35.888 80-80 80s-80-35.888-80-80 35.888-80 80-80 80 35.888 80 80z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </FontAwesomeIcon>
                      <span>
                        ended
                      </span>
                    </span>
                  </StyledComponent>
                </styled.span>
              </StatusDot>
            </div>
            <div
              className="col"
            >
              <div
                className="label"
              >
                upcoming
              </div>
              <StatusDot
                status="upcoming"
              >
                <styled.span
                  className="status-dot upcoming"
                >
                  <StyledComponent
                    className="status-dot upcoming"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bdVaJa",
                          "isStatic": false,
                          "lastClassName": "GNxtY",
                          "rules": Array [
                            "
    & > span { 
        display: none;
    }

    &.",
                            "ongoing",
                            " {
        color: ",
                            "#4caf50",
                            ";
    }
    
    &.",
                            "ended",
                            " {
        color: ",
                            "#ff2717",
                            ";
    }
    
    &.",
                            "upcoming",
                            " {
        color: ",
                            "#ffc107",
                            ";
    }
    
    @media screen and (max-width: ",
                            "25rem",
                            ") {
        & > span { 
            display: inline-block;
            margin-left: ",
                            "0.3125rem",
                            ";
            line-height: 1rem;
        }
    }    
",
                          ],
                        },
                        "displayName": "styled.span",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bdVaJa",
                        "target": "span",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <span
                      className="sc-bdVaJa GNxtY status-dot upcoming"
                    >
                      <FontAwesomeIcon
                        border={false}
                        className=""
                        fixedWidth={false}
                        flip={null}
                        icon={
                          Object {
                            "icon": Array [
                              512,
                              512,
                              Array [],
                              "f192",
                              "M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm80 248c0 44.112-35.888 80-80 80s-80-35.888-80-80 35.888-80 80-80 80 35.888 80 80z",
                            ],
                            "iconName": "dot-circle",
                            "prefix": "fas",
                          }
                        }
                        inverse={false}
                        listItem={false}
                        mask={null}
                        pull={null}
                        pulse={false}
                        rotation={null}
                        size={null}
                        spin={false}
                        swapOpacity={false}
                        symbol={false}
                        title=""
                        transform={null}
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-dot-circle fa-w-16 "
                          data-icon="dot-circle"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          style={Object {}}
                          viewBox="0 0 512 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm80 248c0 44.112-35.888 80-80 80s-80-35.888-80-80 35.888-80 80-80 80 35.888 80 80z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </FontAwesomeIcon>
                      <span>
                        upcoming
                      </span>
                    </span>
                  </StyledComponent>
                </styled.span>
              </StatusDot>
            </div>
          </div>
        </div>
      </div>
    </StyledComponent>
  </styled.div>
  <styled.div
    className="campaign-table"
    dataLabels={
      Object {
        "budget": "Budget (USD)",
        "endDate": "End Date",
        "name": "Name",
        "startDate": "Start Date",
        "status": "Status",
      }
    }
  >
    <StyledComponent
      className="campaign-table"
      dataLabels={
        Object {
          "budget": "Budget (USD)",
          "endDate": "End Date",
          "name": "Name",
          "startDate": "Start Date",
          "status": "Status",
        }
      }
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": false,
            "lastClassName": "jhCusW",
            "rules": Array [
              "
    width: 100%;
    padding: 0 ",
              "1rem",
              " ",
              "1rem",
              ";
    box-sizing: border-box;
    
    & > table {
        border: 0;
        width: 100%;
        border-collapse: collapse;
        
        & td {
            padding: ",
              "0.625rem",
              ";
            border-top: ",
              "1px solid #ddd",
              ";
            border-bottom: ",
              "1px solid #ddd",
              ";
            text-align: center;
        }
        
        & > thead {
            & tr {
                background-color: ",
              "#595caf",
              ";
                
                & td {
                    color: ",
              "#fff",
              ";
                    font-weight: 700;
                }
            }
        }
        
        & > tbody {
            & tr {
                &:nth-child(odd) {
                    background-color: ",
              "#f6f6f6",
              ";
                }
                
                &:not(.no-result):hover {
                    background-color: ",
              "#eee",
              ";
                }
            }
        }
        
        & a {
            text-decoration: underline;
            cursor: pointer;
        }
        
        & .highlight {
            margin-bottom: ",
              "0.625rem",
              ";
            color: ",
              "#ff2717",
              ";
            font-weight: 700;
        }
    }
    
    @media screen and (max-width: ",
              "25rem",
              ") {
        padding: 0 ",
              "1rem",
              " calc(2 * ",
              "1rem",
              ");
        
        & > table {
            & td {
                display: block;
                padding: ",
              "0.3125rem",
              " 0;
                border: 0;
                text-align: left;
                
                ",
              [Function],
              "
            }
            
            & > thead {
                display: none;
            }
            
            & > tbody {
                & tr {
                    display: block;
                    padding: ",
              "1rem",
              ";
                    border: ",
              "1px solid #ddd",
              ";
                }
            }
        }
    }    
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-htpNat jhCusW campaign-table"
      >
        <table>
          <thead>
            <tr>
              <td
                key="name"
              >
                Name
              </td>
              <td
                key="startDate"
              >
                Start Date
              </td>
              <td
                key="endDate"
              >
                End Date
              </td>
              <td
                key="budget"
              >
                Budget (USD)
              </td>
              <td
                key="status"
              >
                Status
              </td>
            </tr>
          </thead>
          <tbody>
            <tr
              className="no-result"
            >
              <td
                colSpan={5}
              >
                <div
                  className="highlight"
                >
                  No campaigns available.
                </div>
                Click 
                <a
                  href="#"
                  onClick={[Function]}
                >
                  here
                </a>
                 to learn how to add new campaigns via Web Development tools.
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </StyledComponent>
  </styled.div>
</CampaignList>
`;
